internal/api/mail.go

```diff
@@ -269,24 +271,39 @@ func (a *API) sendConfirmation(r *http.Request, tx *storage.Connection, u *model
 	referrerURL := utilities.GetReferrer(r, config)
 	externalURL := getExternalHost(ctx)
 	var err error
-	if err := validateSentWithinFrequencyLimit(u.ConfirmationSentAt, maxFrequency); err != nil {
+	if err := validateSentWithinFrequencyLimit(user.ConfirmationSentAt, maxFrequency); err != nil {
 		return err
 	}
-	oldToken := u.ConfirmationToken
+	oldToken := user.ConfirmationToken
 	otp, err := crypto.GenerateOtp(otpLength)
 	if err != nil {
 		// OTP generation must succeeed
 		panic(err)
 	}
-	token := crypto.GenerateTokenHash(u.GetEmail(), otp)
-	u.ConfirmationToken = addFlowPrefixToToken(token, flowType)
+	token := crypto.GenerateTokenHash(user.GetEmail(), otp)
+	user.ConfirmationToken = addFlowPrefixToToken(token, flowType)
 	now := time.Now()
-	if err := mailer.ConfirmationMail(r, u, otp, referrerURL, externalURL); err != nil {
-		u.ConfirmationToken = oldToken
+	if config.Hook.CustomEmailProvider.Enabled {
+		input := hooks.CustomEmailProviderInput{
+			UserID:          user.ID,
+			Email:           user.Email.String(),
+			OTP:             otp,
+			EmailActionType: mail.SignupVerification,
+			UserMetaData:    user.UserMetaData,
+			RedirectTo:      referrerURL,
+		}
```

Comment:
I would maybe scope the email data one level and the user data another level.

Let's provide users with the full `User` object and under `email_data` the OTP, action types, etc.